trl/trainer/online_dpo_trainer.py

```diff
@@ -1,590 +1,444 @@
     get_reward,
     prepare_deepspeed,
-    print_rich_table,
-    truncate_response,
+    trl_sanitze_kwargs_for_tagging,
+    truncate_right,
 )
-from .online_dpo_config import OnlineDPOConfig
-from .utils import trl_sanitze_kwargs_for_tagging
 
 
-INVALID_LOGPROB = 1.0
+if is_apex_available():
+    from apex import amp
+
+
+if is_sagemaker_mp_enabled():
+    from smdistributed.modelparallel import __version__ as SMP_VERSION
+
+    IS_SAGEMAKER_MP_POST_1_10 = version.parse(SMP_VERSION) >= version.parse("1.10")
+
+else:
+    IS_SAGEMAKER_MP_POST_1_10 = False
+
+logger = logging.get_logger(__name__)
 
 
 class OnlineDPOTrainer(Trainer):
     _tag_names = ["trl", "online-dpo"]
 
     def __init__(
         self,
-        config: OnlineDPOConfig,
-        tokenizer: PreTrainedTokenizer,
-        model: nn.Module,
-        ref_model: nn.Module,
-        reward_model: nn.Module,
-        train_dataset: Dataset,
-        data_collator: Optional[DataCollatorWithPadding] = None,
-        eval_dataset: Optional[Union[Dataset, Dict[str, Dataset]]] = None,
-        # less commonly used
-        optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
+        model: Union[PreTrainedModel, nn.Module] = None,
+        ref_model: Union[PreTrainedModel, nn.Module] = None,
+        reward_model: Optional[nn.Module] = None,
+        judge: Optional[BasePairwiseJudge] = None,
+        args: Optional[OnlineDPOConfig] = None,
+        data_collator: Optional[DataCollator] = None,
+        train_dataset: Optional[Union[Dataset, IterableDataset, "datasets.Dataset"]] = None,
+        eval_dataset: Optional[Union[Dataset, Dict[str, Dataset], "datasets.Dataset"]] = None,
+        tokenizer: Optional[PreTrainedTokenizerBase] = None,
+        model_init: Optional[Callable[[], PreTrainedModel]] = None,
+        compute_metrics: Optional[Callable[[EvalPrediction], Dict]] = None,
         callbacks: Optional[List[TrainerCallback]] = None,
+        optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
+        preprocess_logits_for_metrics: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
     ) -> None:
-        self.args = config
-        args = config
-        self.tokenizer = tokenizer
+        self.ref_model = ref_model
 
-        # disable `pad_token_id` and `eos_token_id` because we just want to
-        model.generation_config.eos_token_id = None
-        # generate tokens without truncation / padding
-        model.generation_config.pad_token_id = None
+        if reward_model is not None and judge is not None:
+            warnings.warn(
+                "Both `reward_model` and `judge` are provided. Please choose provide only one of them. "
+                "Ignoring `judge` and using `reward_model`."
+            )
+        elif reward_model is None and judge is None:
+            raise ValueError("Either `reward_model` or `judge` must be provided.")
 
```

Comment:
Note for me: add a NotImplementedError if judge is not None. Let's work on integrating judge in another PR.